drivers/accel/adxl313/iio_adxl313.c

```diff
@@ -0,0 +1,1200 @@
+{
+	int32_t val;
+	uint32_t multiplier = ADXL312_OFFSET_SCALE_FACTOR;
+	struct adxl313_iio_dev *iio_adxl313;
+	struct adxl313_dev *adxl313;
+
+	if (!dev)
+		return -EINVAL;
+
+	iio_adxl313 = (struct adxl313_iio_dev *)dev;
+
+	if (!iio_adxl313->adxl313_dev)
+		return -EINVAL;
+
+	adxl313 = iio_adxl313->adxl313_dev;
+
+	switch (adxl313->dev_type) {
+	case ID_ADXL312:
+		multiplier = ADXL312_OFFSET_SCALE_FACTOR;
+		break;
+	case ID_ADXL313:
+		multiplier = ADXL313_OFFSET_SCALE_FACTOR;
+		break;
+	case ID_ADXL314:
+		multiplier = ADXL314_OFFSET_SCALE_FACTOR;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	switch (channel->type) {
+	case IIO_ACCEL:
+		switch (channel->address) {
+		case ACCEL_AXIS_X:
+			val = adxl313->x_offset_raw * multiplier;
+			return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
+		case ACCEL_AXIS_Y:
+			val = adxl313->y_offset_raw * multiplier;
+			return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
+		case ACCEL_AXIS_Z:
+			val = adxl313->z_offset_raw * multiplier;
+			return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
+
+		default:
+			return -EINVAL;
+		}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+/***************************************************************************//**
+ * @brief Handles the write request for calibbias attribute.
+ *
+ * @param dev     - The iio device structure.
+ * @param buf	  - Command buffer to be filled with the data to be written.
+ * @param len     - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv    - Command attribute id.
+ *
+ * @return ret    - Result of the writing procedure.
+*******************************************************************************/
+static int adxl313_iio_write_calibbias(void *dev, char *buf, uint32_t len,
+				       const struct iio_ch_info *channel, intptr_t priv)
+{
+	int32_t ret;
+	int32_t val;
+	enum adxl313_op_mode op_mode;
+	struct adxl313_iio_dev *iio_adxl313;
+	struct adxl313_dev *adxl313;
+
+	if (!dev)
+		return -EINVAL;
+
+	iio_adxl313 = (struct adxl313_iio_dev *)dev;
+
+	if (!iio_adxl313->adxl313_dev)
+		return -EINVAL;
+
+	adxl313 = iio_adxl313->adxl313_dev;
+
+
+
+	if (channel->type == IIO_ACCEL) {
+		iio_parse_value(buf, IIO_VAL_INT, &val, NULL);
+
+		// The device has to be in standby mode in order to be able to
+		// modify the offset.
+		// Obtain the current op mode
+		op_mode = adxl313->op_mode;
+		// Set the device in standby mode
+		ret = adxl313_set_op_mode(adxl313, ADXL313_STDBY);
+		if (ret)
+			return ret;
+
+		// Set the offset on the requested axis
+		switch (channel->address) {
+		case ACCEL_AXIS_X:
+			ret = adxl313_set_offset(adxl313, val, ADXL313_X_AXIS);
+			break;
+		case ACCEL_AXIS_Y:
+			ret = adxl313_set_offset(adxl313, val, ADXL313_Y_AXIS);
+			break;
+		case ACCEL_AXIS_Z:
+			ret = adxl313_set_offset(adxl313, val, ADXL313_Z_AXIS);
+			break;
+		default:
+			ret = -EINVAL;
+			break;
+		}
+
+		if (ret) {
+			adxl313_set_op_mode(adxl313, op_mode);
```

Comment:
is this function call supposed to be here?

Reply:
If adxl313_set_offset fails, the adxl313_iio_write_calibbias function should return the error code generated by adxl313_set_offset, and not by the adxl313_set_op_mode(which could be executed sucessfully). However, the chip was set to standby mode for performing the setting, even if the setting action had failed. This function call tries to set the device to the state previous to standby (could have been MEASURE).